# Ejercicio 4

> Por: Mapachana

**Dado tu número n de 8 cifras de la lista del ejercicio 2.**

In [156]:
n = 77770081
n

77770081

**i) Facroriza n-1 aplicando el métodoΡ de Polard. ¿Cúantas iteraciones necesitas?**

In [157]:
import math

In [158]:
def f(x):
    return pow(x,2)+1

def rho_polar(n, x0, t0):
    x = x0
    y = x
    t = t0
    i = 0
    g= '-'
    print("Paso: {}, x: {}, y: {}, g: {}".format(i,x,y,g))
    while i < t:
        i = i+1
        x = pow(f(x),1,n)
        y = pow(f(f(y)),1,n)
        g = math.gcd(x-y, n)
        print("Paso: {}, x: {}, y: {}, g: {}".format(i,x,y,g))
        if 1 < g and g < n:
            print("g: {}".format(g))
            print("Iteraciones necesarias {}".format(i))
            
            return g, i
    
    print("No hay divisores con t iteraciones")
    return "No hay divisores con t iteraciones"

In [159]:
rho_polar(7429, 1, 100)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 2957, g: 1
Paso: 4, x: 677, y: 6890, g: 19
g: 19
Iteraciones necesarias 4


(19, 4)

In [160]:
rho_polar(391, 1, 100)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 286, g: 1
Paso: 3, x: 26, y: 220, g: 1
Paso: 4, x: 286, y: 243, g: 1
Paso: 5, x: 78, y: 82, g: 1
Paso: 6, x: 220, y: 220, g: 391
Paso: 7, x: 308, y: 243, g: 1
Paso: 8, x: 243, y: 82, g: 23
g: 23
Iteraciones necesarias 8


(23, 8)

Comenzamos aplicando el metodo $\rho$ de Polard a nuestro número $n$, que al ser par es claramente compuesto.

In [161]:
m = n-1
factores = []
itotal = 0
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i
print(m)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
25923360


Como la descomposición es $77770080=3\cdot 25923360$ el cofactor es claramente compuesto al ser par.

In [162]:
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i
print(m)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 3
g: 3
Iteraciones necesarias 1
8641120


Como la descomposición es $25923360=3\cdot 8641120$ el cofactor es claramente compuesto al ser par.

In [163]:
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i
print(m)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 32
g: 32
Iteraciones necesarias 2
270035


Como la descomposición es $8641120=32\cdot 270035$ el cofactor es claramente compuesto al ser múltiplo de 5.

In [164]:
g, i = rho_polar(m, 1, 100)
factores.append(g)
m = m//g
itotal = itotal + i
print(m)

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 221631, g: 5
g: 5
Iteraciones necesarias 3
54007


In [165]:
pow(2, 54007-1, 54007)

29823

Como la descomposición es $270035=5\cdot 54007$ el cofactor es compuesto ya que $2^{54007-1}\equiv 29823\mod 54007\not\equiv 1\mod 54007$.

In [166]:
g, i = rho_polar(m, 1, 100)
factores.append(g)
itotal = itotal + i
m = m//g

Paso: 0, x: 1, y: 1, g: -
Paso: 1, x: 2, y: 5, g: 1
Paso: 2, x: 5, y: 677, g: 1
Paso: 3, x: 26, y: 5603, g: 1
Paso: 4, x: 677, y: 11539, g: 1
Paso: 5, x: 26274, y: 30672, g: 1
Paso: 6, x: 5603, y: 12599, g: 53
g: 53
Iteraciones necesarias 6


In [167]:
factores.append(m)
print(factores)
print(itotal)

[3, 3, 32, 5, 53, 1019]
13


La descomposición de este número es $54007=53*1019$, donde mirando en la lista vemos que tanto 53 como 1019 son primos, luego la descomposición en factores de nuestro número es:
$n-1=3^2\cdot 2^5\cdot 5 \cdot 53 \cdot 1019$ y el total de iteraciones es $13$.

In [168]:
aux = 1
for elem in factores:
    aux = elem*aux

if aux == n-1:
    print("Bien descompuesto")
else:
    print("dep")

Bien descompuesto


**ii) Si es necesario aplica recursivamente Lucas-Lehmer para certificar factores primos de n-1 mayores de 4 cifras.**

No ha resultado ningún factor mayor de 4 cifras, por lo que no es necesario.

**iii) Aplica Lucas-Lehmer para encontrar un certificado de primalidad de n.**

**NOTA: Debes encontrar el natural más pequeño cuya clase sea primitiva.**

In [169]:
def lucas_lehmer(n, primos):
    fermat = False
    for a in range (1, 20):
        if pow(a, n-1, n) == 1:
            fermat = True
            print(a)
            print("test fermat: {}".format(pow(a, n-1, n)))
            for num in primos:
                print("Para num: {}".format(pow(a, (n-1)//num, n)))

In [170]:
lucas_lehmer(591601 ,[2, 3, 5, 17, 29])

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 1
Para num: 359699
Para num: 134030
Para num: 391816
Para num: 251513
3
test fermat: 1
Para num: 1
Para num: 359699
Para num: 134030
Para num: 551078
Para num: 296851
4
test fermat: 1
Para num: 1
Para num: 231901
Para num: 76535
Para num: 501558
Para num: 77441
5
test fermat: 1
Para num: 1
Para num: 359699
Para num: 164724
Para num: 318079
Para num: 165787
6
test fermat: 1
Para num: 1
Para num: 231901
Para num: 76535
Para num: 419471
Para num: 64560
7
test fermat: 1
Para num: 591600
Para num: 359699
Para num: 1
Para num: 1
Para num: 524350
8
test fermat: 1
Para num: 1
Para num: 1
Para num: 216311
Para num: 429148
Para num: 138510
9
test fermat: 1
Para num: 1
Para num: 231901
Para num: 76535
Para num: 420754
Para num: 364049
10
test fermat: 1
Para num: 1
Para num: 231901
Para num: 1
Para num: 1
Para num: 364049
11
test fermat: 1
Para num: 591600
Para num: 1
Para num: 164724
Para num: 

In [171]:
primos = [2, 3, 5, 53, 1019]

lucas_lehmer(n, primos)

1
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 1
2
test fermat: 1
Para num: 1
Para num: 19635892
Para num: 1
Para num: 66308477
Para num: 51905151
3
test fermat: 1
Para num: 1
Para num: 58134188
Para num: 77424475
Para num: 38633269
Para num: 30239188
4
test fermat: 1
Para num: 1
Para num: 58134188
Para num: 1
Para num: 898507
Para num: 35425890
5
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 32102390
Para num: 58752468
6
test fermat: 1
Para num: 1
Para num: 1
Para num: 77424475
Para num: 46375981
Para num: 1440970
7
test fermat: 1
Para num: 1
Para num: 19635892
Para num: 26472647
Para num: 23657361
Para num: 17963547
8
test fermat: 1
Para num: 1
Para num: 1
Para num: 1
Para num: 60470873
Para num: 18085434
9
test fermat: 1
Para num: 1
Para num: 19635892
Para num: 66432901
Para num: 64599758
Para num: 10633980
10
test fermat: 1
Para num: 1
Para num: 19635892
Para num: 1
Para num: 70122749
Para num: 77143975
11
test fermat: 1
Para num: 1
Para n

Tenemos que $a=17$ es un elemento primitivo para $n=77770081$ porque $17^{n-1}\equiv 1\mod n$ y $17^{\frac{n-1}{p}}\not\equiv 1\mod n $ para $p\in\{2,3,5,53,1019\}$ pues:

$$17^{\frac{n-1}{2}}\equiv 77770080\mod n$$
$$17^{\frac{n-1}{3}}\equiv 58134188\mod n$$
$$17^{\frac{n-1}{5}}\equiv 66432901\mod n$$
$$17^{\frac{n-1}{53}}\equiv 68065795\mod n$$
$$17^{\frac{n-1}{1019}}\equiv 65224721\mod n$$

In [172]:
pow(17, 77770080//1019, 77770081)

65224721